### PR TITLE
Provide the XonshImportEventLoader with all the functions that pkg_re…

### LINKED ISSUE #3607: Xonsh crashes on start, importing prompt_toolkit -> wcwidth fails
Xonsh crashes on startup with a "NotImplementedError" coming from pkg_resources during import of prompt_toolkit -> wcwidth.
The error does not occur with `xonsh --no-rc`, *however* using `source .config/xonsh/rc.xsh` works *fine*. 
Using print statements I have also checked that the rc actually is fully loaded before the crash.

The error started happening after fedora updated wcwidth from 0.1.9 to 0.2.3.
As a temporary workaround, manually downgrading to 0.1.9 seems to restore working function to xonsh.
Another workaround, keeping the newer version, is to place `from wcwidth import wcwidth` in xonshrc (thus skipping xonsh's custom module loading code, I assume).

## xonfig
This is the xonfig when using `xonsh --no-rc` followed by `source .config/xonsh/rc.xsh`.
<details>

```
$ xonfig                                                                       
+------------------+---------------------+
| xonsh            | 0.9.17.dev2         |
| Git SHA          | 97a921c7            |
| Commit Date      | Apr 8 14:48:23 2020 |
| Python           | 3.8.3               |
| PLY              | 3.11                |
| have readline    | True                |
| prompt toolkit   | 2.0.10              |
| shell type       | prompt_toolkit2     |
| pygments         | 2.4.2               |
| on posix         | True                |
| on linux         | True                |
| distro           | fedora              |
| on darwin        | False               |
| on windows       | False               |
| on cygwin        | False               |
| on msys2         | False               |
| is superuser     | False               |
| default encoding | utf-8               |
| xonsh encoding   | utf-8               |
| encoding errors  | surrogateescape     |
+------------------+---------------------+
```

</details>

## Traceback 

<details>

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/xonsh/main.py", line 404, in main
    args = premain(argv)
  File "/usr/lib/python3.8/site-packages/xonsh/main.py", line 357, in premain
    env = start_services(shell_kwargs, args)
  File "/usr/lib/python3.8/site-packages/xonsh/main.py", line 306, in start_services
    builtins.__xonsh__.shell = Shell(execer=execer, **shell_kwargs)
  File "/usr/lib/python3.8/site-packages/xonsh/shell.py", line 181, in __init__
    elif not ptk_above_min_supported():
  File "/usr/lib/python3.8/site-packages/xonsh/platform.py", line 175, in ptk_above_min_supported
    return ptk_version_info()[:2] >= minimum_required_ptk_version
  File "/usr/lib/python3.8/site-packages/xonsh/platform.py", line 167, in ptk_version_info
    return tuple(int(x) for x in ptk_version().strip("<>+-=.").split("."))
  File "/usr/lib/python3.8/site-packages/xonsh/platform.py", line 156, in ptk_version
    import prompt_toolkit
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/__init__.py", line 18, in <module>
    from .application import Application
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/__init__.py", line 3, in <module>
    from .application import Application
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 13, in <module>
    from prompt_toolkit.buffer import Buffer
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/buffer.py", line 17, in <module>
    from .application.current import get_app
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/current.py", line 5, in <module>
    from prompt_toolkit.eventloop.context import TaskLocal, TaskLocalNotSetError
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/eventloop/__init__.py", line 3, in <module>
    from .async_generator import (
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/eventloop/async_generator.py", line 24, in <module>
    from .coroutine import From, Return
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/eventloop/coroutine.py", line 5, in <module>
    from prompt_toolkit.eventloop.defaults import get_event_loop
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/eventloop/defaults.py", line 5, in <module>
    from prompt_toolkit.utils import is_windows
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/utils.py", line 14, in <module>
    from wcwidth import wcwidth
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/usr/lib/python3.8/site-packages/xonsh/imphooks.py", line 294, in exec_module
    rtn = self.loader.exec_module(module)
  File "/usr/lib/python3.8/site-packages/wcwidth/__init__.py", line 25, in <module>
    __version__ = _get_package_version()
  File "/usr/lib/python3.8/site-packages/wcwidth/wcwidth.py", line 408, in _get_package_version
    pkg_resources.resource_string(
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 1156, in resource_string
    return get_provider(package_or_requirement).get_resource_string(
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 1401, in get_resource_string
    return self._get(self._fn(self.module_path, resource_name))
  File "/usr/lib/python3.8/site-packages/pkg_resources/__init__.py", line 1570, in _get
    raise NotImplementedError(
NotImplementedError: Can't perform this operation for loaders without 'get_data()'
Xonsh encountered an issue during launch
Failback to /bin/sh
```

</details>

### PR DESCRIPTION
**Pull request by Achimh3011**
…sources would expect.

It is not clear why they are now missed when it worked for years without them. Comments very welcome, also on the methods thate are redirected - maybe not all are needed / maybe some are missing.

Fixes #3607

### CODE REVIEW

**Comment by Achimh3011** [comment]
That is definitely a better approach than implementing all these wrapper methods. I'll try it out and modify the PR if it works.

My expectation would be that other Python implementations will work with the new approach as well. Maybe it is harder to optimize, but the semantics should hold.

File: xonsh/imphooks.py [hunk 1]
```diff
@@ -295,13 +313,10 @@ def exec_module(self, module):
...
         return rtn
 
-    def load_module(self, fullname):
-        """Legacy module loading, provided for backwards compatibility."""
-        return self.loader.load_module(fullname)
-
-    def module_repr(self, module):
-        """Legacy module repr, provided for backwards compatibility."""
-        return self.loader.module_repr(module)
+    def __getattribute__(self, name):
```
Comment (astronouth7303) [inline xonsh/imphooks.py:316]: Why this over `__getattr__`?

Comment (laloch) [reply xonsh/imphooks.py:316]: Oops! That's my fault. @astronouth7303 is right - `__getattr__` is the one we need.

Comment (Achimh3011) [reply xonsh/imphooks.py:316]: Argh, right! It worked and I did not think twice...

**Comment by scopatz** [comment]
Hi @Achimh3011 - What is the purpose of this exactly? I haven't read the issue this is purporting to fix yet, but it seems like we should avoid `pkg_resources` at all costs.  On some platforms, `pkg_resources` adds minutes to the startup time of xonsh

**Comment by Achimh3011** [comment]
Well, as you can see in #3607, on some platforms xonsh does not start at all, because dependencies use `pkg_resources` and the overridden import mechanics of xonsh do not support that. 

Why this has become a problem only recently (and only on some platforms), I don't know. Prior to this PR, there were two "compat" methods present, now a larger number of methods is supported. 

As this is all based on dynamic lookup, the PR does not alter the (non-)dependency of xonsh, but ensures xonsh works with dependencies using `pkg_resources`.

File: xonsh/imphooks.py [hunk 2]
```diff
@@ -270,6 +270,24 @@ def find_spec(self, fullname, path, target=None):
         return spec
 
 
+_XIEVL_WRAPPED_ATTRIBUTES = [
```
Comment (scopatz) [inline xonsh/imphooks.py:273]: This should be a frozenset

**Comment by scopatz** [comment]
Right, so we go out of our way to make sure that our critical dependencies do not use pkg_resources. I guess I am OK with this, subject to the frozenset comment

**Comment by scopatz** [comment]
Also, this needs a news entry :wink:

**Comment by scopatz** [comment]
Thanks!

**Pull request merged**